### PR TITLE
IGNITE-15945 Class inheritance and effectively final fields

### PR DESCRIPTION
**Pull request by rpuch**
https://issues.apache.org/jira/browse/IGNITE-15945

### CODE REVIEW

File: modules/network/src/main/java/org/apache/ignite/internal/network/serialization/ClassDescriptor.java [hunk 1]
```diff
@@ -70,6 +71,23 @@
...
 
+    /**
+     * Size of the nulls bitmap for the described class; it is equal to the number of nullable (i.e. non-primitive)
+     * fields that have a type known upfront.
+     *
+     * @see #fieldNullsBitmapIndices
+     * @see #isValueTypeKnownUpfront()
+     */
+    private final int fieldNullsBitmapSize;
+    /**
```
Comment (SammyVimes) [inline modules/network/src/main/java/org/apache/ignite/internal/network/serialization/ClassDescriptor.java:82]: Missing newline? Or is it optional

File: modules/network/src/main/java/org/apache/ignite/internal/network/serialization/marshal/DefaultUserObjectMarshaller.java [hunk 2]
```diff
@@ -274,30 +195,56 @@ private boolean mayHaveObjectIdentity(ClassDescriptor descriptor) {
...
         ProtocolMarshalling.writeObjectId(objectId, output);
 
-        writeObject(describedObject.object, describedObject.descriptor, output, context);
+        writeObject(object, descriptor, output, context);
     }
 
-    private void marshalValue(DescribedObject describedObject, DataOutputStream output, MarshallingContext context)
-            throws IOException, MarshalException {
-        writeDescriptorId(describedObject.descriptor, output);
+    private boolean concreteTypeIsKnownUpfront(@Nullable Class<?> declaredClass) {
```
Comment (SammyVimes) [inline modules/network/src/main/java/org/apache/ignite/internal/network/serialization/marshal/DefaultUserObjectMarshaller.java:221]:
```suggestion
    private boolean specificTypeIsKnownUpfront(@Nullable Class<?> declaredClass) {
```
Concrete is a building material :)

**Review by SammyVimes — APPROVED** [review]
LGTM

File: modules/network/src/main/java/org/apache/ignite/internal/network/serialization/ClassDescriptor.java (hunk 1 above)
Comment (rpuch) [reply modules/network/src/main/java/org/apache/ignite/internal/network/serialization/ClassDescriptor.java:83]: Added the newline

File: modules/network/src/main/java/org/apache/ignite/internal/network/serialization/marshal/DefaultUserObjectMarshaller.java (hunk 2 above)
Comment (rpuch) [reply modules/network/src/main/java/org/apache/ignite/internal/network/serialization/marshal/DefaultUserObjectMarshaller.java]: Changed the terminology to mention runtime type instead (it seems really suitable in this case).

**Pull request closed without merging**